database: impl trait function bodies for `heed` & `redb`

hinto-janai opened the pull request:
## Part 10
https://github.com/Cuprate/cuprate/pull/81 <- Previous

- Implements `cuprate_database` trait function bodies for the `heed` and `redb` backends

Boog900 commented on database/src/backend/heed/database.rs:
```diff
@@ -26,72 +26,127 @@ use crate::{
+where
+    Range: RangeBounds<T::Key> + RangeBounds<&'tx T::Key> + 'tx,
+{
+    /// TODO
+    struct Iter<'tx, T: Table> {
+        /// TODO
+        iter: heed::RoRange<'tx, StorableHeed<T::Key>, StorableHeed<T::Value>>,
+    }
+
+    // TODO
+    impl<'tx, T: Table> Iterator for Iter<'tx, T> {
+        type Item = Result<&'tx T::Value, RuntimeError>;
+        fn next(&mut self) -> Option<Self::Item> {
+            // TODO
+            self.iter
+                .next()
+                .map(|result| result.map(|value| value.1).map_err(RuntimeError::from))
+        }
+    }
+
+    // TODO
+    match db.range(tx_ro, &range) {
+        Ok(iter) => Ok(Iter::<'_, T> { iter }),
+        Err(e) => Err(RuntimeError::from(e)),
+    }
```
I don't think you need a custom iterator here:

```suggestion
    Ok(db.range(tx_ro, &range)?.map(|res| Ok(res?.1)))
```

Boog900 commented on database/src/backend/heed/database.rs:
```diff
@@ -26,72 +26,127 @@ use crate::{
+    ) -> Result<impl Iterator<Item = Result<impl Borrow<T::Value> + 'a, RuntimeError>>, RuntimeError>
     where
-        <T as Table>::Value: 'a,
+        Range: RangeBounds<T::Key> + RangeBounds<&'a T::Key> + 'a,
```
just make this RangeBounds<T::Key> IMO, this will only be used on keys with simple types (u64)

hinto-janai replied:
This is due to https://docs.rs/redb/latest/redb/trait.ReadableTable.html#tymethod.range.

- `heed` wants `Range: RangeBounds<T::Key>`
- `redb` wants `Range: RangeBounds<impl Borrow<T::Key>>`

The current `redb` code is `&T::Key`. If we switch our key to `Cow<'_, T::Key>` too, our range has to `impl Borrow<Cow<'_, T::Key>>`.

I guess the input would be:
```rust
let range = Cow::<u64>::Owned(1)..Cow::<u64>::Owned(10);
```

This does create a incompatibility when switching backends since `heed` wants a `u64` and `redb` wants a `Cow<'_, u64>`. We could switch it so `heed` also uses `Cow` internally I guess...?

Boog900 replied:
doesn't T::Key impl Borrow for it self? https://doc.rust-lang.org/std/borrow/trait.Borrow.html#impl-Borrow%3CT%3E-for-T

hinto-janai replied:
To be more clear:
- `heed` wants `Range: RangeBounds<T::Key>` (`heed` sees `&'a T` as the key)
- `redb` wants `Range: RangeBounds<Cow<'a, T::Key>> + 'a` (`redb` sees `Cow<'_, T>` as the key)

We could switch `heed` so it sees `Cow<'_, T>` instead of `&'a T` as the key type, but means we're wrapping all key/values in it during serialization (not bad, still gets serialized as T). It means our ranges will be like the above though...

Or we can make `redb` accept `RangeBounds<T::Key>` then in `redb`'s specific `get_range()` we somehow transform:
```rust
RangeBounds<T::Key>
// ...into...
RangeBounds<Cow<'?, T::Key>>
```
is this possible?

hinto-janai commented:
@Boog900 ready. `cuprate_database` can finally insert/retrieve values...

Additions:
- `trait ValueGuard` to abstract over `redb::AccessGuard` and `Cow<'_, Key>`
- `trait ToOwnedDebug` to allow `Storable` types to be `ToOwned`, `Debug`, and their `Owned` versions to also be `Debug`
- `trait EnvInner` to abstract over `heed` requiring the caller to hold a guard object while using the environment as it uses locks internally 

Changes:
- `redb`'s key/value type from `Table::Key` -> `Cow<'_, Table::Key>` due to unaligned bytes causing UB and/or panicking when casting. Although, any `T` where alignment is `1` should be casted cheaply

Misc:
- Separated `config.rs` -> `config/`
- Added `cuprate_database::config::Backend` (does nothing for now)
- Main traits are documented with invariants
- Made changes to CI

Boog900 commented on database/src/backend/heed/transaction.rs:
```diff
@@ -9,31 +11,26 @@ use crate::{
-    /// TODO
     fn commit(self) -> Result<(), RuntimeError> {
-        todo!()
+        self.commit().map_err(Into::into)
```
and here:
```suggestion
        Ok(self.commit()?)
```

Boog900 commented on database/src/backend/heed/transaction.rs:
```diff
@@ -9,31 +11,26 @@ use crate::{
-    /// TODO
     fn commit(self) -> Result<(), RuntimeError> {
-        todo!()
+        self.commit().map_err(Into::into)
```
```suggestion
        Ok(self.commit()?)
```

Boog900 commented on database/src/backend/redb/database.rs:
```diff
@@ -1,62 +1,201 @@
+    db: &'a impl redb::ReadableTable<StorableRedb<T::Key>, StorableRedb<T::Value>>,
+    key: &'a T::Key,
+) -> Result<impl ValueGuard<T::Value> + 'a, RuntimeError> {
+    match db.get(Cow::Borrowed(key)) {
+        Ok(Some(redb_guard)) => Ok(redb_guard),
+        Ok(None) => Err(RuntimeError::KeyNotFound),
+        Err(e) => Err(RuntimeError::from(e)),
+    }
```
```suggestion
    db.get(Cow::Borrowed(key))?.ok_or(RuntimeError::KeyNotFound)
```

Boog900 commented on database/src/backend/redb/database.rs:
```diff
@@ -1,62 +1,201 @@
+    where
+        K: ToOwnedDebug,
+    {
+        // FIXME: we must match and re-wrap since this function wants `&Cow` not `Cow`.
+        fn start_bound(&self) -> Bound<&Cow<'a, K>> {
+            match &self.start_bound {
+                Bound::Included(t) => Bound::Included(t),
+                Bound::Excluded(t) => Bound::Excluded(t),
+                Bound::Unbounded => Bound::Unbounded,
+            }
+        }
+
+        // FIXME: we must match and re-wrap since this function wants `&Cow` not `Cow`.
+        fn end_bound(&self) -> Bound<&Cow<'a, K>> {
+            match &self.end_bound {
+                Bound::Included(t) => Bound::Included(t),
+                Bound::Excluded(t) => Bound::Excluded(t),
+                Bound::Unbounded => Bound::Unbounded,
+            }
+        }
```
```suggestion
        fn start_bound(&self) -> Bound<&Cow<'a, K>> {
            self.start_bound.as_ref()
        }

        fn end_bound(&self) -> Bound<&Cow<'a, K>> {
            self.end_bound.as_ref()
        }
```

Boog900 commented on database/src/backend/redb/database.rs:
```diff
@@ -1,62 +1,201 @@
 
-    fn get_range<'a>(
+    #[inline]
+    #[allow(clippy::unnecessary_wraps, clippy::trait_duplication_in_bounds)]
```
I don't think this is needed anymore (`allow`)

Boog900 commented on database/src/backend/redb/database.rs:
```diff
@@ -1,62 +1,201 @@
 
-    fn get_range<'a>(
+    #[inline]
+    #[allow(clippy::unnecessary_wraps, clippy::trait_duplication_in_bounds)]
```
same here

Boog900 commented on database/src/backend/redb/env.rs:
```diff
@@ -61,55 +67,124 @@ impl Env for ConcreteEnv {
+
+        // TODO: Create all tables when schema is done.
+
+        // TestTable
+        let table: redb::TableDefinition<
+            'static,
+            StorableRedb<<TestTable as Table>::Key>,
+            StorableRedb<<TestTable as Table>::Value>,
+        > = redb::TableDefinition::new(TestTable::NAME);
+        tx_rw.open_table(table)?;
+
+        // TestTable2
+        let table: redb::TableDefinition<
+            'static,
+            StorableRedb<<TestTable2 as Table>::Key>,
+            StorableRedb<<TestTable2 as Table>::Value>,
+        > = redb::TableDefinition::new(TestTable2::NAME);
+        tx_rw.open_table(table)?;
+
```
same here with heed about seperating this into a function at some point so we don't have to repeate this for each table

Boog900 commented on .github/workflows/ci.yml:
```diff
@@ -96,22 +96,16 @@ jobs:
         update: true
         install: mingw-w64-x86_64-toolchain mingw-w64-x86_64-boost msys2-runtime-devel git mingw-w64-x86_64-cmake mingw-w64-x86_64-ninja
 
-    - name: Update Rust (UNIX)
-      if: matrix.os != 'windows-latest'
-      run: rustup update
-
-    - name: Switch target (Windows)
-      if: matrix.os == 'windows-latest'
-      run: rustup toolchain install stable-x86_64-pc-windows-gnu -c clippy && rustup set default-host x86_64-pc-windows-gnu && rustup default stable-x86_64-pc-windows-gnu
-
```
hmm weird, I don't think these lines caused that error. 

Maybe this got cached? and thats why `windows-gnu` was already installed

hinto-janai replied:
Should this be added back? CI works without it.

Boog900 replied:
If CI works then this can stay removed

Boog900 commented on database/README.md:
```diff
@@ -66,21 +66,23 @@ Note that `lib.rs/mod.rs` files are purely for re-exporting/visibility/lints, an
+| `to_owned_debug.rs` | Borrowed/owned data abstraction; `trait ToOwnedDebug`
+| `transaction.rs`    | Database transaction abstraction; `trait TxR{o,w}`
+| `types.rs`          | Database table schema types
+| `value_guard.rs`    | Database table schema types
```
was purposed missed for `value_guard`?

Boog900 commented on database/src/backend/heed/database.rs:
```diff
@@ -26,72 +32,113 @@ use crate::{
+    tx_ro: &'a heed::RoTxn<'_>,
+    key: &T::Key,
+) -> Result<impl ValueGuard<T::Value> + 'a, RuntimeError> {
+    match db.get(tx_ro, key) {
+        Ok(Some(cow)) => Ok(Cow::Borrowed(cow)),
+        Ok(None) => Err(RuntimeError::KeyNotFound),
+        Err(e) => Err(e.into()),
+    }
```
```suggestion
    db.get(tx_ro, key)?
        .map(Cow::Borrowed)
        .ok_or(RuntimeError::KeyNotFound)
```

Boog900 reviewed (changes requested):
Looks good! left some comments

hinto-janai commented on database/src/backend/heed/env.rs:
```diff
@@ -77,45 +94,133 @@ impl Drop for ConcreteEnv {
+
+        // TODO: Create all tables when schema is done.
+
+        /// Function that creates the tables based off the passed `T: Table`.
+        fn create_table<T: Table>(
+            env: &heed::Env,
+            tx_rw: &mut heed::RwTxn<'_>,
+        ) -> Result<(), InitError> {
+            DatabaseOpenOptions::new(env)
+                .name(<T as Table>::NAME)
+                .types::<StorableHeed<<T as Table>::Key>, StorableHeed<<T as Table>::Value>>()
+                .create(tx_rw)?;
+            Ok(())
+        }
+
+        create_table::<TestTable>(&env, &mut tx_rw)?;
+        create_table::<TestTable2>(&env, &mut tx_rw)?;
```
`heed`'s `create_table()` (local function in `Env::open`)

hinto-janai commented on database/src/backend/redb/env.rs:
```diff
@@ -61,55 +67,123 @@ impl Env for ConcreteEnv {
+        fn create_table<T: Table>(tx_rw: &redb::WriteTransaction<'_>) -> Result<(), InitError> {
+            let table: redb::TableDefinition<
+                'static,
+                StorableRedb<<T as Table>::Key>,
+                StorableRedb<<T as Table>::Value>,
+            > = redb::TableDefinition::new(<T as Table>::NAME);
+
+            // `redb` creates tables on open if not already created.
+            tx_rw.open_table(table)?;
+            Ok(())
+        }
 
-    #[cold]
-    #[inline(never)] // called once in [`Env::open`]?`
-    fn create_tables<T: Table>(&self, tx_rw: &mut Self::TxRw<'_>) -> Result<(), RuntimeError> {
-        todo!()
+        create_table::<TestTable>(&tx_rw)?;
+        create_table::<TestTable2>(&tx_rw)?;
```
`redb`'s `create_table()` (local function in `Env::open`)

hinto-janai reviewed (commented):
Made changes and added the `create_table()` functions.

Boog900 commented on database/src/database.rs:
```diff
@@ -1,54 +1,74 @@
+    /// if a particular key in the `range` does not exist,
+    /// [`RuntimeError::KeyNotFound`] wrapped in [`Err`] will be returned
+    /// from the iterator.
+    #[allow(clippy::trait_duplication_in_bounds)]
```
```suggestion
```

Boog900 reviewed (changes requested):
just one clippy allow thats isn't needed anymore

hinto-janai commented:
Fixed, also removed some documentation about our backend's seeing all types as `[u8]` for serde, which isn't true anymore.

They see `Storable{Heed,Redb}` as the implementer, then either `&T` or `Cow<'_, T>` as the type.
